lib/stellar/transaction_builder.rb

```diff
@@ -0,0 +1,108 @@
+module Stellar
+  class TransactionBuilder
+
+    attr_reader :source_account, :sequence_number, :base_fee, :time_bounds, :time_bounds, :memo, :operations
+
+    def initialize(
+      source_account:, 
+      sequence_number:, 
+      base_fee: 100, 
+      time_bounds: nil, 
+      memo: nil
+    )
+      raise ArgumentError, "bad :source_account" unless source_account.is_a?(Stellar::KeyPair)
+      raise ArgumentError, "bad :sequence_number" unless sequence_number.is_a?(Integer) && sequence_number > 0
+      raise ArgumentError, "bad :time_bounds" unless time_bounds.is_a?(Stellar::TimeBounds) || time_bounds.nil?
+      raise ArgumentError, "bad :base_fee" unless base_fee.is_a?(Integer) && base_fee >= 100
+
+      @source_account = source_account
+      @sequence_number = sequence_number
+      @base_fee = base_fee
+      @time_bounds = time_bounds
+      @memo = self.make_memo(memo)
+      @operations = Array.new
+    end
+
+    def build
+      if @time_bounds.nil?
+        raise "TransactionBuilder.time_bounds must be set during initialization or by calling set_timeout"
+      elsif !@time_bounds.min_time.is_a?(Integer) or !@time_bounds.max_time.is_a?(Integer)
+        raise "TimeBounds.min_time and max_time must be Integers"
+      elsif @time_bounds.max_time != 0 and @time_bounds.min_time > @time_bounds.max_time
+        raise "Timebounds.max_time must be greater than min_time"
+      elsif @time_bounds.max_time != 0 and @time_bounds.max_time < Time.now.to_i
+        raise "Timebounds.max_time must be in the future"
+      end
+      tx = Stellar::Transaction.new(
+        source_account: @source_account.account_id,
+        fee: @base_fee * @operations.length,
+        seq_num: @sequence_number,
+        time_bounds: @time_bounds,
+        memo: @memo,
+        operations: @operations,
+        ext: Stellar::Transaction::Ext.new(0)
+      )
+      @sequence_number += 1
```

Comment:
This is how the [JS SDK](https://stellar.github.io/js-stellar-sdk/node_modules_stellar-base_src_transaction_builder.js.html#line182) is implemented as well.

Reply:
😱 Oh dear.

Technically its not quite the same because this gets and increments the sequence number on the source object, which might be more or less shocking depending on how we look at it. Technically the transaction builder in the JS SDK has the source account as a reference and calling `build` builds the next transaction for that referenced source account. Even though I'd argue that this is still 😱, it's explainable through that referenced relationship.

In this PR the transaction has a sequence number and calling build increments the transactions sequence number.

Either way I guess this point to this being necessary.

Reply:
Yeah, JS did this first as a convenience. We copied it in Go SDK for least surprise (other SDKs do this also). It's a side effect which can be surprising. Also it's annoying when you actually don't want the seqnum incremented. This [discussion](https://github.com/stellar/go/issues/1259) is worth a read.

Reply:
So whats the takeaway? Should we remove the line that increments the sequence number and require users call `set_sequence_number` if they're building multiple transactions with the same builder?

Reply:
Didn't refresh before making my comment. I'll keep the increment. Users can always use `set_sequence_number` if they want different functionality.